internal/fleet/integration_policy/secrets.go

```diff
@@ -0,0 +1,165 @@
+package integration_policy
+
+import (
+	"context"
+	"encoding/json"
+
+	fleetapi "github.com/elastic/terraform-provider-elasticstack/generated/fleet"
+	"github.com/elastic/terraform-provider-elasticstack/internal/utils"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+)
+
+// The secret store is a map of policy secret reference IDs to the
+// original value at time of creation. By replacing the ref when
+// marshaling the state back to Terraform, we can prevent resource
+// drift.
+type secretStore map[string]any
+
+// newSecretStore creates a new secretStore from the resource privateData.
+func newSecretStore(ctx context.Context, private privateData) (store secretStore, diags diag.Diagnostics) {
+	bytes, diags := private.GetKey(ctx, "secrets")
+	if diags != nil {
+		return
+	}
+	if bytes == nil {
+		bytes = []byte("{}")
+	}
+
+	err := json.Unmarshal(bytes, &store)
+	if err != nil {
+		diags.AddError("could not unmarshal secret store", err.Error())
+		return
+	}
+
+	return
+}
+
+// Save marshals the secretStore back to the provider.
+func (s secretStore) Save(ctx context.Context, private privateData) (diags diag.Diagnostics) {
+	bytes, err := json.Marshal(s)
+	if err != nil {
+		diags.AddError("could not marshal secret store", err.Error())
+		return
+	}
+
+	return private.SetKey(ctx, "secrets", bytes)
+}
+
+// pruneRefsFromResponse removes any saved secret refs not present in the API response.
+func pruneRefsFromResponse(resp *fleetapi.PackagePolicy, secrets secretStore) {
+	refs := make(map[string]any)
+	for _, r := range utils.Deref(resp.SecretReferences) {
+		refs[*r.Id] = nil
+	}
+
+	for id := range secrets {
+		if _, ok := refs[id]; !ok {
+			delete(secrets, id)
+		}
+	}
+}
```

Comment:
I'd probably prefer returning a new `secretStore` containing only the refs in the response rather than manipulating the existing map. WDYT? 

```suggestion
func (secrets secretStore) pruneRefsFromResponse(resp *fleetapi.PackagePolicy) secretStore {
	refs := make(map[string]any)
	for _, r := range utils.Deref(resp.SecretReferences) {
		refs[*r.Id] = nil
	}

	newSecrets := secretStore{}
	for id, secret := range secrets {
		if _, ok := refs[id]; ok {
			newSecrets[id] = secret
		}
	}
}
```

Reply:
since the entire process is only called once each CRUD method, I'll simplify it by pulling in the entire load/save mess into the handle methods. The resp will be available there as well, so we can merge the secrets filter directly into the contructor at that point. 

Ultimately, all read/create/update will see, is `diags = handleReqRespSecrets(ctx, body, policy, resp.Private)` and the rest will all be encapsulated.